### PR TITLE
Allow mending of faults for already installed parts [RDY]

### LINKED ISSUE #16438: Engine faults can't be viewed/mended without removing the engine

### PR DESCRIPTION
**Pull request by mugling**
Fixes #16438

### CODE REVIEW

**Comment by Coolthulhu** [comment]
Not directly related to this PR, but it would be cool if new features that restrict something came with a mod to disable them, at least until they stop being experimental.

**Comment by mugling** [comment]
Sure, in this case the new functionality converts already broken (greyed-out) engines to ones with faults however your general point holds true.

Once we introduce spontaneous vehicle faults I'll add a 'no vehicle maintenance mod' or equivalent

**Comment by Coolthulhu** [comment]
Needs astyle

**Comment by Coolthulhu** [comment]
* Time to mend isn't displayed
* Still hard to test - have to find a vehicle that gets the fault during mapgen
* Mend is not bound in vehicle menu

File: data/json/faults.json [hunk 1]
```diff
@@ -15,6 +16,7 @@
     "type": "fault",
     "name": "glow plugs",
     "description": "Required to start the vehicle in low ambient tmperatures.",
```
Comment (Coolthulhu) [inline data/json/faults.json]: "tmperatures"

**Comment by mugling** [comment]
> Mend is not bound in vehicle menu

Does 8523889 correctly fix this?

**Comment by mugling** [comment]
Ready

### Significant features
* Fault repairs become a long action (`ACT_MEND_ITEM`)
* Faults can be toggled in debug mode (`DEBUG_HS`)
* Faults can be disabled via a mod (`no_faults`)

### Implementation details
* Extend `item_location` to support returning the location `type()` and `position()`
* Cleanup of the `MINUTES`, `HOURS` and `DAYS` macros to be `constexpr` functions
* Implement `calendar::print_duration` for formatted display of durations (in turns)
* Add `MAX_FIELD_DENSITY` and start replacing hardcoded literals with it

### Balance
This PR adds automotive filters providing further non-critical faults. Penalties are chosen to lead to the possibility of emergent behavior. Increased smoke (among other effects) from a faulty air filter isn't too problematic until you also damage the vehicles muffler and need then need to rev the engine hard to make an escape.

I want to follow this pattern with faults producing additional challenges for a player as opposed to completely crippling an engine. Spare parts should be a relatively valuable resource and I want to build upon the improvised repair manual to include more makeshift parts.

**Comment by Coolthulhu** [comment]
Conflicts with #16487

I'll test #16487 first, since it fixes a major bug.

**Comment by Coolthulhu** [comment]
That conflict doesn't look trivial - this PR seems to allow using car parts as components (?)

**Comment by mugling** [comment]
> That conflict doesn't look trivial - this PR seems to allow using car parts as components (?)

Have resolved conflict. The special case is so that you can obtain an `item_location` pointing to the base item of a vehicle

**Comment by Coolthulhu** [comment]
> The special case is so that you can obtain an item_location pointing to the base item of a vehicle which allows mending of items in the inventory and installed in vehicles to share the same code path.

What is it used for? Sounds like it could result in counterintuitive things, like ripping off a roof to install it in other place.

**Comment by mugling** [comment]
> What is it used for?

Providing an argument to `player::mend_item`

> Sounds like it could result in counterintuitive things, like ripping off a roof to install it in other place.

No. `item_location` is just a generic handle to an item plus some accessory functions describing it's location. The functionality you describe is provided by the separate `visitable` interface

File: src/activity_handlers.cpp [hunk 2]
```diff
@@ -1501,6 +1502,51 @@ void activity_handlers::repair_item_finish( player_activity *act, player *p )
...
+    for( const auto& e : reqs.get_components() ) {
+        p->consume_items( e );
+    }
+    for( const auto& e : reqs.get_tools() ) {
+        p->consume_tools( e );
+    }
+    p->invalidate_crafting_inventory();
+
+    target->faults.erase( *f );
+    add_msg( m_good, _( "You successfuly mended the %s." ), target->tname().c_str() );
```
Comment (illi-kun) [inline src/activity_handlers.cpp]: "successfuly" -- typo (one missed "l").

File: src/player.cpp [hunk 3]
```diff
@@ -9928,6 +9928,24 @@ bool player::dispose_item( item& obj, const std::string& prompt )
 
 void player::mend_item( item_location&& obj, bool interactive )
 {
+    if( g->u.has_trait( "DEBUG_HS" ) ) {
+        uimenu menu( true, "Toggle which fault?" );
```
Comment (illi-kun) [inline src/player.cpp]: I'd prefer to see this string translatable (even with taking into account that it's available only with debug trait).

File: src/player.cpp [hunk 4]
```diff
@@ -9928,6 +9928,24 @@ bool player::dispose_item( item& obj, const std::string& prompt )
...
+        std::vector<std::pair<fault_id, bool>> opts;
+        for( const auto& f : obj->faults_potential() ) {
+            opts.emplace_back( f, obj->faults.count( f ) );
+            menu.addentry( -1, true, -1, string_format( "%s %s", opts.back().second ? "Mend" : "Break", f.obj().name().c_str() ) );
```
Comment (illi-kun) [inline src/player.cpp]: and these too.

File: src/calendar.cpp [hunk 5]
```diff
@@ -260,7 +260,40 @@ float calendar::sunlight() const
     }
 }
 
+std::string calendar::print_duration( int turns )
```
Comment (BevapDin) [inline src/calendar.cpp]: This function looks very similar to `calendar::textify_period`, maybe those should be combined?

There is also similar code to print the time for constructions (`construction::get_time_string`) and for crafting (`recipe::print_time`) that could be replaced by this function.

File: src/field.cpp [hunk 6]
```diff
@@ -2496,19 +2496,10 @@ field_id field_entry::setFieldType(const field_id new_field_id){
...
-    } else {
-        density = new_density;
-    }
-
-    return density;
-
+int field_entry::setFieldDensity(const int new_density)
+{
+    is_alive = new_density > 1;
+    return std::max( std::min( new_density, MAX_FIELD_DENSITY ), 1 );
```
Comment (BevapDin) [inline src/field.cpp]: `this->density` is not changed by the function anymore. The function name is a lie.

File: src/activity_handlers.cpp [hunk 7]
```diff
@@ -1501,6 +1502,51 @@ void activity_handlers::repair_item_finish( player_activity *act, player *p )
     act->moves_left = actor->move_cost;
 }
 
+void activity_handlers::mend_item_finish( player_activity *act, player *p )
+{
+    item_location target;
+    switch( static_cast<item_location::type>( act->index ) ) {
```
Comment (BevapDin) [inline src/activity_handlers.cpp]: Exposing the location types seems bad. `item_location` should be handle this via serialize/deserialize functions. For example with forwarding the player_activity to them.

Comment (mugling) [inline src/activity_handlers.cpp]: Could you expand upon the idea. I've wanted to tackle the interaction between `player_activity` and `item_location` for some time as there's a lot of interesting mechanics and fixes we can implement once this is done.

Comment (BevapDin) [inline src/activity_handlers.cpp]: Your current solution exposes implementation details (the `type`, position) and makes modifying/extending the class more complicated. `item_location` itself should be responsible for that task.

We need some way to store an `item_location` in a `player_activity`. Ideally we should be able to store several of them (for example to repair item X using item Y), a `vector<item_location>` seems appropriate.

I suggest this: change `item_location` to deriving from `JsonSerializer` and `JsonDeserializer`, see other class like npc/player/item/...). Store item locations in a vector in the activity and serialized the vector the same way the other two vectors are serialized.

Serializing is relatively simple (see savegame_json.cpp), as is deserialization. You may need to add error checking and special handling for the item pointer. The pointer itself can not be stored, you need to store some abstract reference (e.g. the index in the item list) and recalculated the pointer after deserialization. Deserailization may result in a null-location, but this needs to be handled by the activity anyway as the item may get lost during the activity.

This allows extending / modifying `item_location` (even adding new types, like in the inventory of pets) without any changes outside of that class / that cpp file. It's also usable in other parts of the game and allows storing several locations in the `player_activity`.

In short: a user of the `item_location` shouldn't need to know about the internals, they only need to know about the serialize/deserialize functions.

File: src/calendar.cpp (hunk 5 above)
Comment (mugling) [inline src/calendar.cpp]: Agreed, but separate PR

File: src/activity_handlers.cpp (hunk 7 above)
Comment (mugling) [inline src/activity_handlers.cpp]: This could work but is non-trivial. I'd like to push this PR for now as it includes a number of fixes and greatly improves item faults. I'm going to start work on your suggestion in a separate PR

**Comment by mugling** [comment]
All issues fixed. `item_location` serialization will be done in a separate PR

File: data/json/faults.json [hunk 8]
```diff
@@ -36,17 +39,43 @@
...
   {
+    "id": "fault_engine_filter_air",
+    "type": "fault",
+    "name": "air filter",
+    "description": "An expired filter reduces fuel efficiency and increases smoke production.",
+    "time": 30000,
+    "skills": [ [ "mechanics", 1 ] ],
+    "requirements": {
+      "qualities": [ { "id": "SCREW", "level": 1 } ],
+      "components": [ [ [ "filter", 1 ], [ "filter_makeshift", 1 ] ] ]
```
Comment (kevingranade) [inline data/json/faults.json]: Based on the filter description, it's a fuel filter, you need something totally different for air filtering.

File: src/calendar.h [hunk 9]
```diff
@@ -3,10 +3,20 @@
...
-#define HOURS(x)   ((x) * 600)
-#define DAYS(x)    ((x) * 14400)
+constexpr int MINUTES( int n )
+{
+    return n * 10;
+}
+
+constexpr int HOURS( int n )
+{
+    return n * 600;
```
Comment (codemime) [inline src/calendar.h]: A minor thing, but anyways: `n * MINUTES( 60 )` looks nicer.

File: src/calendar.h [hunk 10]
```diff
@@ -3,10 +3,20 @@
...
+}
+
+constexpr int DAYS( int n )
+{
+    return n * 14400;
```
Comment (codemime) [inline src/calendar.h]: Same as above: `n * HOURS( 24 )` would be nicer.

File: src/activity_handlers.cpp (hunk 7 above)
Comment (kevingranade) [inline src/activity_handlers.cpp]: The rest of the PR checks out, but I'm not ok merging this part of the change, it's really hard to put the genie back in the bottle once it's out, especially since in this case it involves writing the player_activity to a savefile.  We need to get that part out of the way up front so we don't have to patch in conversion code.

I'm working on this feature right now, it should be ready today or tomorrow.

Comment (mugling) [inline src/activity_handlers.cpp]: > I'm working on this feature right now, it should be ready today or tomorrow.

Good stuff. I'll rebase this when your ready

File: data/json/faults.json (hunk 8 above)
Comment (mugling) [inline data/json/faults.json]: Hmmn, I was trying to represent that the material both are comprised of is analogous but I suppose we do need differing types. Perhaps the improvised repairs manual should allow you to improvise one type from the other?

File: src/activity_handlers.cpp (hunk 7 above)
Comment (mugling) [inline src/activity_handlers.cpp]: Any progress and/or do you want to collaborate on this. I'm presuming your taking the approach of trying to serialize `item_location`?

Comment (mugling) [inline src/activity_handlers.cpp]: As the activity isn't suspendable it won't get written out to the save so future extension would not be so problematic. I'm concious this PR fixes some otherwise significant deficiencies in the faults system.

**Comment by Coolthulhu** [comment]
Non-trivial conflict here. I think with the inheritance code. I don't know how hard is it to fix, so I'll leave this one for now.

**Comment by mugling** [comment]
@Coolthulhu happy if you want to merge this prior to the inheritance code and I can then rebase?

File: src/item_factory.cpp [hunk 11]
```diff
@@ -91,6 +91,10 @@ void Item_factory::finalize() {
     for( auto& e : m_templates ) {
         itype& obj = *e.second;
 
+        if( obj.engine && item_options.count( "no_faults" ) ) {
```
Comment (Coolthulhu) [inline src/item_factory.cpp]: I'm getting an error here:
```
src/item_factory.cpp:95:27: error: use of undeclared identifier 'item_options'
```

File: data/mods/No_Faults/modinfo.json [hunk 12]
```diff
@@ -0,0 +1,16 @@
+[
+  {
+    "type": "MOD_INFO",
+    "mod-type": "SUPPLEMENTAL",
+    "ident": "no_faults",
+    "name": "Perfect Reliability",
+    "author": "Mugling",
+    "description": "All items are perfectly reliable and never spawn with faults or develop them during use",
```
Comment (Coolthulhu) [inline data/mods/No_Faults/modinfo.json]: Misleading - it suggests guns don't get damaged during use and possibly even that melee weapons and armors don't get damaged.

File: data/json/faults.json [hunk 13]
```diff
@@ -14,7 +15,8 @@
     "id": "fault_engine_glow_plug",
     "type": "fault",
     "name": "glow plugs",
-    "description": "Required to start the vehicle in low ambient tmperatures.",
+    "description": "Required to start the vehicle in low ambient temperatures.",
```
Comment (Coolthulhu) [inline data/json/faults.json]: Required? Last time I checked it was just helpful, not necessary.

**Comment by Coolthulhu** [comment]
Mending itself works.

The mod to disable faults doesn't work and in fact breaks the build, as it depends on some other PR. I think it actually conflicts with the PR it depends on.

The debug fault toggling menu is bugged pretty hard, but it allows toggling some faults.

File: data/mods/No_Faults/modinfo.json (hunk 12 above)
Comment (mugling) [inline data/mods/No_Faults/modinfo.json]: I'm not so sure. It states that items don't develop faults, not that they are never damaged.

Comment (mugling) [inline data/mods/No_Faults/modinfo.json]: That said we could adjust the description to limit it to vehicle parts until other faults are added

**Comment by mugling** [comment]
@coolthulhu all other problems fixed. I think your compilation issue is due to a bad merge as I can compile (from clean) without any problems.

**Comment by Coolthulhu** [comment]
> I think your compilation issue is due to a bad merge as I can compile (from clean) without any problems. 

Are you sure you didn't merge some extra branch in or something like that?
Here, `grep "item_options" -ri src` doesn't return even a single line on the master branch.

**Comment by mugling** [comment]
Yes, see above edit. It's due to #16577. I'm resolving it now

**Comment by mugling** [comment]
1b1b744 should resolve that

**Comment by Coolthulhu** [comment]
The perfect reliability mod gets a json error on load due to dependency on `ITEM_OPTIONS` field.

For now the perfect reliability could be removed if re-implementing it without the dependency would be too much work, but it would be nice to have as soon as possible.

**Comment by mugling** [comment]
That fix should be trivial, let me see what I can do

File: src/activity_handlers.cpp (hunk 7 above)
Comment (Coolthulhu) [inline src/activity_handlers.cpp]: While this is a rather serious problem, I'm going to merge this PR unless someone tells me not to because current mending system is a common source of legitimate complaints and frustration.

This regression should be fixed ASAP and should be treated like a bug, though.

> As the activity isn't suspendable it won't get written out to the save

Not true: an autosave in the middle of the activity will write the activity with hardcoded switch values. Future updates will need to handle that.

Comment (mugling) [inline src/activity_handlers.cpp]: Ok. Can you open an issue after merging and I'll start work on serialization of `item_location` next as that would have a lot of extra benefits

**Comment by mugling** [comment]
See 5f911f1

File: src/activity_handlers.cpp (hunk 7 above)
Comment (Coolthulhu) [inline src/activity_handlers.cpp]: The more I think about it the worse it sounds.

Could we just reuse the idea from crafting for now? That is, instead of tracking item location, just consume the first-best item that solves the problem.
It will stop being correct when replacement parts become actual parts of the items, but it will be a while until that happens.

**Comment by Coolthulhu** [comment]
How about instead of having the hack with item locations, disable mending non-vehicle-mounted engines for now? It's not like anyone would ever want to fix an engine they wouldn't want to use on a vehicle.

**Comment by mugling** [comment]
That might cause some player confusion and I'd prefer avoiding changing the interface further? The worst that can happen in the current scenario is that if the game was to autosave during a repair and then the player subsequently loaded/scummed that save the repair would fail. I'm working on a PR right now for `item_location` but it will need some extensive testing so I'd rather push this PR which fixes some major flaws with that one small problem and tackle serialization separately.

**Comment by mugling** [comment]
See also #16791

File: src/item_location.cpp [hunk 14]
```diff
@@ -342,6 +372,16 @@ const item *item_location::operator->() const
...
 }
 
+item_location::type item_location::where() const
+{
+    return ptr ? ptr->where() : type::invalid;
+}
+
+tripoint item_location::position() const
+{
+    return ptr ? ptr->position() : tripoint();
```
Comment (Coolthulhu) [inline src/item_location.cpp]: Should be `tripoint_min` - just `tripoint` will be `(0, 0, 0)`, which is in reality bubble

Comment (mugling) [inline src/item_location.cpp]: Noted

Comment (mugling) [inline src/item_location.cpp]: Fixed in 8a89fa8
